### PR TITLE
TSan CI

### PR DESCRIPTION
**Pull request by bjacob**
This adds a TSan CI bot under `buildkite/presubmit` and postsubmit.

Example run: https://buildkite.com/iree/presubmit/builds/261 -> `Executing build-and-test-tsan` -> https://buildkite.com/iree/unregistered/builds/43

### CODE REVIEW

**Comment by bjacob** [comment]
https://buildkite.com/iree/unregistered/builds/39#01813fc7-82d2-423c-a135-c8136c484190 failed with

```
[2022-06-07T20:16:41Z] /usr/bin/ld: final link failed: No space left on device
```

While it is expected that the TSan build would use more disk space at link time, if only because it uses the system linker, it seems like the build machine should be able to accomodate that?

**Comment by GMNGeoffrey** [comment]
> https://buildkite.com/iree/unregistered/builds/39#01813fc7-82d2-423c-a135-c8136c484190 failed with
> 
> ```
> [2022-06-07T20:16:41Z] /usr/bin/ld: final link failed: No space left on device
> ```
> 
> While it is expected that the TSan build would use more disk space at link time, if only because it uses the system linker, it seems like the build machine should be able to accomodate that?
> 
> EDIT: actually it is failing at linking the compiler itself, so this is purely the impact of TSan instrumentation.

We don't yet have a productionized agent setup, but agreed this shouldn't happen. For now, I'll manually resize the disk

**Comment by bjacob** [comment]
Got a commit locally adding `IREE_ENABLE_LLD` as I saw that `rebuild.sh` does that. Let me know when the disk is resized and i'll push again...

**Comment by bjacob** [comment]
@GMNGeoffrey now we are hitting this error:

```
[2022-06-07T21:07:44Z] 2022-06-07 21:07:44 ERROR  Error uploading artifact "build-tsan-01813fe4-99f3-40ee-8054-812479444f45.tgz": File size (5974852070 bytes) exceeds the maximum supported by Buildkite's default artifact storage (5Gb). Alternative artifact storage options may support larger files.
```

Is the fix to increase this limit (we're not far above the existing limit, 5.9G vs 5.0G) or would it help to actually perform the testing in the same action as the build, not in a separate dependent action --- would that remove the need for the 'upload' that is too big here?

**Comment by bjacob** [comment]
Updated this PR to do build AND test in one step. This successfully avoids the previous issue and we are now actually running TSan CTests! Yay! But: our TSan greenness regressed since it was established two months ago. We have multiple data races to fix first, before we're TSan green again and could enable this CI. I'm going to file issues and refer to them from the parent issue for TSan CI, https://github.com/google/iree/issues/8969.

**Comment by bjacob** [comment]
Update: this is all green and ready for review!

File: build_tools/cmake/build_and_test_tsan.sh [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+# Disable actually running GPU tests. This tends to yield TSan reports that are
+# specific to one's particular GPU driver and therefore hard to reproduce across
+# machines and often un-actionable anyway.
+# See e.g. https://github.com/google/iree/issues/9393
+export IREE_VULKAN_DISABLE=1
+export IREE_CUDA_DISABLE=1
+
+# Honor the "notsan" label on tests.
+export IREE_NOTSAN_DISABLE=1
```
Comment (ScottTodd) [inline build_tools/cmake/build_and_test_tsan.sh:79]: The double negative here is a little confusing, as is the lack of a separator in "notsan".

I can't think of a more clear name for _this particular setting_, so how about we instead have `ctest_all` read an extra list of values for `label_exclude_args`? Then this script could do `export IREE_EXCLUDE_LABELS=notsan`.

We could also split "notsan" into "no-tsan" or "no_tsan", but consistency with other labels is important too.

File: build_tools/bazel/lit_test.bzl [hunk 2]
```diff
@@ -61,6 +61,7 @@ def lit_test(
         data = None,
         visibility = None,
         env = None,
+        timeout = None,
```
Comment (GMNGeoffrey) [inline build_tools/bazel/lit_test.bzl:64]: Seem like you've got some unrelated diffs here. Could you merge from/rebase on main?

File: build_tools/cmake/build_and_test_tsan.sh (hunk 1 above)
Comment (GMNGeoffrey) [reply build_tools/cmake/build_and_test_tsan.sh:79]: Yeah especially because `NOTSAN` is easily parse as "Not San". The main reason the other things use environment variables that create label exclusions is that we also use them as shell variables (`IREE_VULKAN_F16_DISABLE` is the exception here). They're also disabling functionality *within IREE*, which is different than filtering out tests marked with "notsan", I think.

Comment (bjacob) [reply build_tools/cmake/build_and_test_tsan.sh]: Done.

**Comment by bjacob** [comment]
The base PR #9435 is merged, now is a good time to approve this PR 😁

**Review by GMNGeoffrey — APPROVED** [review]
